maven-resolver-synccontext-simple/src/main/java/org/eclipse/aether/named/simple/SimpleSyncContextFactory.java

```
@@ -0,0 +1,368 @@
+import org.eclipse.aether.metadata.Metadata;
+import org.eclipse.aether.spi.synccontext.SyncContextFactory;
+
+@Named
+@Singleton
+public class SimpleSyncContextFactory implements SyncContextFactory {
+
+    public static final String REQUEST_CONTEXT = "request-context";
+    public static final String REQUEST_ACQUIRE = "request-acquire";
+    public static final String REQUEST_CLOSE = "request-close";
+    public static final String RESPONSE_CONTEXT = "response-context";
+    public static final String RESPONSE_ACQUIRE = "response-acquire";
+    public static final String RESPONSE_CLOSE = "response-close";
+
+    Socket socket;
+    DataOutputStream output;
+    DataInputStream input;
+    Thread receiver;
+    AtomicInteger requestId = new AtomicInteger();
+    Map<Integer, CompletableFuture<List<String>>> responses = new ConcurrentHashMap<>();
+
+    @Override
+    public SyncContext newInstance(RepositorySystemSession session, boolean shared) {
+        return new SimpleSyncContext(session, shared);
+    }
+
+    synchronized Socket ensureInitialized() throws IOException {
+        if (socket == null) {
+            socket = createClient();
+            input = new DataInputStream(socket.getInputStream());
+            output = new DataOutputStream(socket.getOutputStream());
+            receiver = new Thread(this::receive);
+            receiver.setDaemon(true);
+            receiver.start();
+        }
+        return socket;
+    }
+
+    synchronized Socket createClient() throws IOException {
+        Path registryFile = Paths.get(System.getProperty("user.home"))
+                .resolve(".m2/sync.bin").toAbsolutePath().normalize();
+        if (!Files.isRegularFile(registryFile)) {
+            if (!Files.isDirectory(registryFile.getParent())) {
+                Files.createDirectories(registryFile.getParent());
+            }
+        }
+
+        try (RandomAccessFile raf = new RandomAccessFile(registryFile.toFile(), "rw")) {
```

Will fail when user home is on NFS.

---

The locking part is quite important to ensure proper behavior.  I think the lock file could be made configurable.  I'll add a configuration property `aether.syncContext.simple.lockFile` so that we can point to a non NFS.  I can also fail fast during in the constructor to try and lock that file.

---

A property is a good idea.